### PR TITLE
Add SO3.exp_map() and SO3.log_map()

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: theseus/geometry/so3.py [hunk 1]
```diff
@@ -78,10 +79,54 @@ def _unit_quaternion_check(quaternion: torch.Tensor):
...
+        scale = torch.where(sel, theta / sth, 1 + sth ** 2 / 6)
+        ret1 *= scale.view(-1, 1)
+        m00 = self.data[:, 0, 0]
+        m11 = self.data[:, 1, 1]
+        m22 = self.data[:, 2, 2]
+        major = torch.logical_and(m11 > m00, m11 > m22) + 2 * torch.logical_and(
+            m22 > m00, m22 > m11
+        )
+        major += torch.arange(0, 3 * major.shape[0], 3, device=major.device)
+        ret2 = self.data.view(-1, 3)[major].clone()
```
Comment (fantaosha) [inline theseus/geometry/so3.py:125]: Is there a more elegant way to get ret2?

File: theseus/geometry/so3.py [hunk 2]
```diff
@@ -78,10 +79,57 @@ def _unit_quaternion_check(quaternion: torch.Tensor):
...
     @staticmethod
     def exp_map(tangent_vector: torch.Tensor) -> LieGroup:
-        raise NotImplementedError
+        if tangent_vector.ndim != 2 or tangent_vector.shape[1] != 3:
+            raise ValueError("Invalid input for SO3.exp_map.")
+        ret = SO3(dtype=tangent_vector.dtype)
+        theta = torch.linalg.norm(tangent_vector, dim=1, keepdim=True).unsqueeze(1)
+        theta2 = theta ** 2
+        sel = theta >= 0.005
+        a = torch.where(sel, theta.cos(), 8 / (4 + theta2) - 1)
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/geometry/so3.py:88]: I will add some comments later today.

File: theseus/geometry/so3.py [hunk 3]
```diff
@@ -78,10 +79,57 @@ def _unit_quaternion_check(quaternion: torch.Tensor):
...
+        sel1 = 1 + cth > 1e-9
+        scale1 = torch.where(
+            theta[sel1] >= 5e-3, theta[sel1] / sth[sel1], 1 + sth[sel1] ** 2 / 6
+        )
+        ret[sel1] *= scale1.view(-1, 1)
+        sel2 = ~sel1
+        m00 = self.data[sel2, 0, 0]
+        m11 = self.data[sel2, 1, 1]
+        m22 = self.data[sel2, 2, 2]
+        major2 = torch.logical_and(m11 > m00, m11 > m22) + 2 * torch.logical_and(
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/geometry/so3.py:124]: Will do, thanks!

File: theseus/geometry/tests/test_so3.py [hunk 4]
```diff
@@ -20,11 +20,70 @@ def _create_random_so3(batch_size, rng):
...
 
 
-# def test_inverse():
-#     rng = torch.Generator()
-#     for batch_size in [1, 20, 100]:
-#         so3 = _create_random_so3(batch_size, rng)
-#         check_inverse(so3)
+def check_SO3_log_map(tangent_vector, group_cls):
+    error = (tangent_vector - group_cls.exp_map(tangent_vector).log_map()).norm(dim=1)
+    error = torch.minimum(error, (error - 2 * np.pi).abs())
```
Comment (luisenp) [inline theseus/geometry/tests/test_so3.py:25]: I just noticed this part. Is this necessary (taking the min with error - 2pi)? Just want to make sure this is not hiding a bug.

File: theseus/geometry/so3.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (luisenp) [reply theseus/geometry/so3.py:89]: I was looking into doing a small refactor in SE2 as mentioned above, and tried this

image

I checked this by running the tactile pushing example, and confirmed that the forward pass is the same. However, I get `NaN` gradients in the backward pass, and I suspect it is due to the use of `where`. 

**Original**
image 
**New**
image



Not sure if there is an easy work around for this, but if not, this means that as long as we keep using torch autograd to generate gradients for these operations, we might hit a similar issue with SE3.

**Comment by fantaosha** [comment]
A possible problem of `SO3.log_map()` is that there are relatively large errors around `pi` for `float32`. This is largely due to the division by zero and seems difficult to handle. Any thoughts @luisenp @mhmukadam ?

File: theseus/geometry/tests/test_so3.py (hunk 4 above)
Comment (luisenp) [reply theseus/geometry/tests/test_so3.py:25]: I think we had similar errors when implementing one of the 2D factors, but in the end we didn't end up needing to handle this special case. Unfortunately, I wasn't able to find the PR where we had this conversation. Maybe @mhmukadam remembers? 

Two reasons why this doesn't happen in the 2D case are that are:

- your tests are more stringent, so that may explain it (btw, I would like to add similar test for 2D)
- in SE2, the approximation is triggered when |cos(theta)| ~ 1, which include -pi, 0, and pi. Maybe we can do something similar here? (where sel2 is defined, I think) If it works, this may help with the issues around `float32`?

Comment (luisenp) [reply theseus/geometry/tests/test_so3.py:25]: Actually, nevermind, I just recalled that the reason we don't trigger similar errors in SE2 is because we are using float64 in the tests, otherwise we have to handle `2pi` manually as well (I just confirmed this by changing to float32).

File: theseus/geometry/so3.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/geometry/so3.py]: Updated the code for so3._log_map_impl

**Comment by fantaosha** [comment]
> A possible problem of `SO3.log_map()` is that there are relatively large errors around `pi` for `float32`. This is largely due to the division by zero and seems difficult to handle. Any thoughts @luisenp @mhmukadam ?

File: theseus/geometry/so3.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/geometry/so3.py:88]: The bug has been fixed. Please refer to #49.

[the remark this answers is not in the document]
Comment (luisenp) [reply theseus/geometry/so3.py:88]: That's great, thanks!
